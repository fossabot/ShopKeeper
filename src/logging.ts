// Copyright (C) 2018 Sleep E-Z USA, Inc. / Evan Darwin
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {colorSuccess} from './theme';

const chalk = require('chalk');

const INDENT_WIDTH = 3;
const INDENT_STAGE_WIDTH = INDENT_WIDTH * 3;

export type UnknownLogData = Array<null|string|number|{}>;
export class ShopifyLogger {
  console: Console;

  constructor() {
    this.console = global.console;
  }

  logStageStart(message: string, stage: number) {
    this.info(`${chalk.bold.red(`Stage ${stage}`)} :: ${message}`);
  }

  logStageMessage(message: string) {
    const indent = ' '.repeat(INDENT_STAGE_WIDTH - 1);
    this.info(indent + `${chalk.cyan('::')} ${message}`);
  }

  logStageItemCount(items: {[type: string]: number}) {
    const itemCount =
        Object.values(items).reduce((a: number, b: number) => (a + b), 0);
    this.logStageMessage(' ');
    this.logStageMessage(chalk.white(
        `${colorSuccess(itemCount)} Items Processed In This Stage:`));

    // Let's quickly sort them for easy readability
    type SortListItem = [string, number];
    const sortedList =
        Object.keys(items)
            .map(
                (itemType):
                    SortListItem => {
                      const count = items[itemType];
                      return [itemType, count];
                    })
            .sort((a: SortListItem, b: SortListItem) => (a[1] - b[1]))
            .reverse();

    sortedList.forEach(([type, count]) => {
      const indent = ' '.repeat(INDENT_STAGE_WIDTH - 1);

      this.info(
          indent +
          `::   ${chalk.white('+')} ${
              chalk.underline.yellow(
                  padStr(type, 11))} : ${chalk.green(count)}`);
    });
  }

  log(data: string, ...args: UnknownLogData) {
    this.console.log(data, ...args);
  }

  info(data: string, ...args: UnknownLogData) {
    this.console.info(chalk.cyan(`[I] ` + data), ...args);
  }

  warn(data: string, ...args: UnknownLogData) {
    this.console.warn(chalk.yellow(`[W] ` + data), ...args);
  }

  error(data: string|Error, ...args: UnknownLogData) {
    let errStr;

    if (data instanceof Error) {
      const stack = (data.stack || '');
      const allButMsg = stack.split('\n').slice(1).join('\n');
      errStr = `${chalk.redBright(data.message.toString())}\n` +
          `${chalk.gray(allButMsg)}`;
    } else {
      errStr = chalk.redBright(data);
    }

    this.console.error(chalk.red(`\n[E] ` + errStr), ...args);
  }
}

export function padStr(content: string, width: number) {
  const left = width - content.length;
  if (left <= 0) return content;

  return content + ' '.repeat(left);
}