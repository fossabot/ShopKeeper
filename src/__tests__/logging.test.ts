import {padStr, ShopifyLogger} from '../logging';
import {UndefinedStackError} from '../testing/data';

describe('Logging Utilities', () => {
  it('padStr()', () => {
    type PadStrTestArr = [string, number, string];
    const test1: PadStrTestArr = ['hello', 10, 'hello     '];
    const test2: PadStrTestArr = ['verylongword', 2, 'verylongword'];
    const test3: PadStrTestArr = ['exactlen', 8, 'exactlen'];

    expect(padStr(test1[0], test1[1])).toEqual(test1[2]);
    expect(padStr(test2[0], test2[1])).toEqual(test2[2]);
    expect(padStr(test3[0], test3[1])).toEqual(test3[2]);
  });

  it('is instantiable', () => {
    const logger = new ShopifyLogger();

    expect(logger).toBeTruthy();
    expect(logger).toBeInstanceOf(ShopifyLogger);
  });

  describe('Common Functions', () => {
    it('log()', () => {
      global.console = {
        ...global.console,
        log: jest.fn(),
        info: jest.fn(),
      } as Console;

      const logger = new ShopifyLogger();
      logger.log('Hello, world - %s', 'log');

      expect(global.console.info).toMatchSnapshot();
      expect(global.console.log).toMatchSnapshot();
    });

    it('info()', () => {
      global.console = {
        ...global.console,
        log: jest.fn(),
        info: jest.fn(),
      } as Console;

      const logger = new ShopifyLogger();
      logger.info('I have some info! - %s', 'info');

      expect(global.console.info).toMatchSnapshot();
      expect(global.console.log).toMatchSnapshot();
    });

    it('warn()', () => {
      global.console = {
        ...global.console,
        log: jest.fn(),
        warn: jest.fn(),
      } as Console;

      const logger = new ShopifyLogger();
      logger.warn('I have a warning! - %s', 'warn');

      expect(global.console.warn).toMatchSnapshot();
      expect(global.console.log).toMatchSnapshot();
    });

    it('error()', () => {
      global.console = {
        ...global.console,
        log: jest.fn(),
        error: jest.fn(),
      } as Console;

      const logger = new ShopifyLogger();
      const customError = new UndefinedStackError('Some error');

      logger.error('I have some info! - %s', 'err');
      logger.error(customError);

      expect(global.console.log).toMatchSnapshot();
      expect(global.console.error).toMatchSnapshot();
      expect((global.console.error as jest.Mock<{}>).mock.calls[1][0])
          .toContain('[E]');
      expect((global.console.error as jest.Mock<{}>).mock.calls[1][0])
          .toContain('Some error');
    });
  });

  describe('Staging', () => {
    it('logStageStart()', () => {
      global.console = {
        ...global.console,
        log: jest.fn(),
        info: jest.fn(),
      } as Console;

      const logger = new ShopifyLogger();
      logger.logStageStart('Hello, world', 0);
      logger.logStageStart('Hello, world', 1);

      // Added to ensure that console.info() is called, not .log()
      expect(global.console.log).toMatchSnapshot();
      expect(global.console.info).toMatchSnapshot();
    });

    it('logStageMessage()', () => {
      global.console = {
        ...global.console,
        log: jest.fn(),
        info: jest.fn(),
      } as Console;

      const logger = new ShopifyLogger();
      logger.logStageMessage('Some more progress');
      logger.logStageMessage('Doing stuff...');

      expect(global.console.log).toMatchSnapshot();
      expect(global.console.info).toMatchSnapshot();
    });

    it('logStageItemCount()', () => {
      global.console = {
        ...global.console,
        log: jest.fn(),
        info: jest.fn(),
      } as Console;

      const logger = new ShopifyLogger();
      logger.logStageItemCount({
        Something: 5,
        Cool: 20,
      });

      expect(global.console.log).toMatchSnapshot();
      expect(global.console.info).toMatchSnapshot();
    });
  });
});