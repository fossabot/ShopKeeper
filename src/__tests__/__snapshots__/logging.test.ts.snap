// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logging Utilities Common Functions error() 1`] = `[MockFunction]`;

exports[`Logging Utilities Common Functions error() 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "
[E] I have some info! - %s",
      "err",
    ],
    Array [
      "
[E] Some error
",
    ],
  ],
}
`;

exports[`Logging Utilities Common Functions info() 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[I] I have some info! - %s",
      "info",
    ],
  ],
}
`;

exports[`Logging Utilities Common Functions info() 2`] = `[MockFunction]`;

exports[`Logging Utilities Common Functions log() 1`] = `[MockFunction]`;

exports[`Logging Utilities Common Functions log() 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Hello, world - %s",
      "log",
    ],
  ],
}
`;

exports[`Logging Utilities Common Functions warn() 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[W] I have a warning! - %s",
      "warn",
    ],
  ],
}
`;

exports[`Logging Utilities Common Functions warn() 2`] = `[MockFunction]`;

exports[`Logging Utilities Staging logStageItemCount() 1`] = `[MockFunction]`;

exports[`Logging Utilities Staging logStageItemCount() 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[I]         ::  ",
    ],
    Array [
      "[I]         :: 25 Items Processed In This Stage:",
    ],
    Array [
      "[I]         ::   + Cool        : 20",
    ],
    Array [
      "[I]         ::   + Something   : 5",
    ],
  ],
}
`;

exports[`Logging Utilities Staging logStageMessage() 1`] = `[MockFunction]`;

exports[`Logging Utilities Staging logStageMessage() 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[I]         :: Some more progress",
    ],
    Array [
      "[I]         :: Doing stuff...",
    ],
  ],
}
`;

exports[`Logging Utilities Staging logStageStart() 1`] = `[MockFunction]`;

exports[`Logging Utilities Staging logStageStart() 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[I] Stage 0 :: Hello, world",
    ],
    Array [
      "[I] Stage 1 :: Hello, world",
    ],
  ],
}
`;
